### DISCUSSION TITLE
Missing  `nanobind::vectorize`

### DISCUSSION THREAD

**Discussion by mwittgen** [wjakob/nanobind#508] [Q&A]
Does nanobind support `vectorize` or  is there a workaround ?

**Comment by wjakob** [comment]
It's missing, and there is no workaround. I think that the implementation of this feature in pybind11 is way overkill (i.e. far too complicated) -- I'm open to PRs reproducing this feature, but they'd have to be considerably simpler.

**Reply by llohse** [reply]
+1

This is the biggest blocker that prevents us from adopting nanobind---we use `py::vectorize` all over the place.
Would it be feasible to implement something similar for functions with 1, 2, or 3 parameters? 

The interface would not necessarily have to be automatically inferred as in pybind11 (just calling `vectorize` on a function pointer). Something like 
```C++
using arg1_t = nb::ndarray<double, nb::device::cpu>;
using arg2_t = nb::ndarray<double, nb::device::cpu>;
using arg3_t = nb::ndarray<double, nb::device::cpu>;

nb::class_<MyClass>(m, "MyClass")
      .def("my_fn", nb::vectorize<arg1_t, arg2_t, arg3_t>(&MyClass::my_func));
```
The crucial part is to check the input shapes for compatibility and loop over all input dimensions. This is really nothing one wants to write wrappers for each time.

**Reply by wjakob** [reply]
This is not on my to do list, but I would be open to external contributions. The important thing would also be to embrace C+17 `if constexpr` to simplify the pybind11 implementation (it is a bit on the complicated end)
